### ISSUE TITLE
version-based annotations for deprecation and stability?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#20911]
### A Python Precedent

For a long time, I have been envious of what Python shows in their docs in terms of the versions of Python in which a standard library function become available or changed significantly. Here is an example from https://docs.python.org/3/library/os.html#os.fsencode :

> os.fsencode(filename)
>
> Encode [path-like](https://docs.python.org/3/glossary.html#term-path-like-object) filename to the [filesystem encoding and error handler](https://docs.python.org/3/glossary.html#term-filesystem-encoding-and-error-handler); return [bytes](https://docs.python.org/3/library/stdtypes.html#bytes) unchanged.
> [fsdecode()](https://docs.python.org/3/library/os.html#os.fsdecode) is the reverse function.
> *New in version 3.2.*
> *Changed in version 3.6:* Support added to accept objects implementing the [os.PathLike](https://docs.python.org/3/library/os.html#os.PathLike) interface.

These "New in version" and "Changed in version" elements are attached to the individual functions and that can be very useful when developing Python code that needs to run with several versions of Python.

### Imagine a World Where...

What if the Chapel programming language supported attributes/annotations for this purpose?

Here is a straw example just to convey the idea.

`getFileSize` was added in 1.12, so it could indicate that:

``` chapel
@versions 1.12..
proc getFileSize(name: string) { }
```

Then we would get a warning if we compile in a way indicating that we want to be able to run in 1.11:
```
$ chpl --min-version 1.11 --max-version 1.29
warning: FileSystem.getFileSize was not available in version 1.11 -- it was new in 1.12
```

So that's nice and helps to catch easy mistakes. (It would not catch *all* problems with compiling with many versions of Chapel -- you would test for that -- but it can catch the errors that are now obvious to the compiler).

But, the idea can help more. Let's take a tricky example where a function's return type changed. Let's say we had this in 1.26:

``` chapel
proc foo() : int { ... }
```

and then in 1.29 we wanted to change the return type to `void`:

``` chapel
proc foo() : void { ... }
```

We could express that with `where` clauses and the new annotations:

``` chapel
@versions 1.26..1..27
@deprecated "this 'foo' is deprecated -- the newer version returns void -- please specify --min-version 1.29 or newer to accept the new behavior and avoid this warning"
proc foo() : int where minVersion < 1.29 { ... }
@versions 1.26..
proc foo() : void where minVersion > 1.29 { ... }
```

(in the above, I am supposing that the `minVersion` symbol is a `config param` connected to `--min-version`).

Now if you compile it with a `--min-version` that is older than 1.29, you would get the deprecation error:

```
$ chpl --min-version 1.26 --max-version 1.29
warning: this 'foo' is deprecated -- the newer version returns void -- please specify --min-version 1.29 or newer to accept the new behavior and avoid this warning
```

But, if you indicate that you want compatibility only with version 1.29, the deprecation error goes away:
```
$ chpl --min-version 1.29 --max-version 1.29
<no deprecation warning for 'foo'>
```

This seems like it might be an interesting alternative as compared to creating `config param`s for each individual such change in behavior.

**Comment by cassella** [comment]
The documentation improvement seems neat!


The second use case seems a little stickier.  How would it interact with user code that's written to be backwards compatible?  I imagine that would be something like

```chapel
if chplVersion < createVersion(1,29)
  var rc = foo();
else
  foo();
```

Would that trigger the deprecation warning with `--min-version 1.26 --max-version 1.29`?

As written, it seems like compiling with `--min-version 1.26` would end up resolving to the `void foo()` no matter what `--max-version` was given?  So when compiling in this mode, the new `foo()` is just not available?

So I think in my example, the `rc = foo()` branch would be folded out, but the `else` branch would still resolve to the deprecated int-returning call?  (Unless the feature proposes to clamp `minVersion` and `maxVersion` in blocks with `chplVersion` restrictions?)

(Changing the return type seems like a confusing example since the call site can't unambiguously call the new one.)


If I understand, the old deprecated versions of the functions (and their implementations) would have to be left in the modules on main for as many releases as you wanted `--min-version` to be able to consider them?


I'm not following the relationship between the described version history of `foo()` and the particular `@version` tags.  If the `void` version was introduced in `1.29`, why does the last annotation have `@versions 1.26..`?  I don't understand what the `1.27` part of the first annotation `1.26..1.27` is doing if the change in behavior was in `1.29`.  Neither where clause matches `minVersion == 1.29` exactly.


Do the `@versions` annotation affect compilation/resolution; or are they only for documentation, and only `minVersion`/`maxVersion` tests control deprecation warnings and resolution?

I take it the intention is that every public function added in say `1.30` will be added with an `@versions 1.30..`?

**Comment by mppf** [comment]
> The documentation improvement seems neat!
> 
> The second use case seems a little stickier. How would it interact with user code that's written to be backwards compatible? I imagine that would be something like
> 
> ```chapel
> if chplVersion < createVersion(1,29)
>   var rc = foo();
> else
>   foo();
> ```
> 
> Would that trigger the deprecation warning with `--min-version 1.26 --max-version 1.29`?

I would expect it would. If the author didn't want the deprecation warning, they would use a newer `min-version`.

> As written, it seems like compiling with `--min-version 1.26` would end up resolving to the `void foo()` no matter what `--max-version` was given? So when compiling in this mode, the new `foo()` is just not available?

Yes, I simplified to get the starting idea across. In reality I think we would need those `where` clauses to ask something about the range of versions we are compiling for.

> If I understand, the old deprecated versions of the functions (and their implementations) would have to be left in the modules on main for as many releases as you wanted `--min-version` to be able to consider them?

That's true but it's not my intention here to tie the deprecation timeline to this feature. As a result maybe we'd want to call it something more non-committal, like `--min-version-deprecation-warnings` and `--max-version-deprecation-warnings`, but that leaves out the use case of identifying a call to a function that didn't exist yet. But maybe that is a separate feature.

> I'm not following the relationship between the described version history of `foo()` and the particular `@version` tags. If the `void` version was introduced in `1.29`, why does the last annotation have `@versions 1.26..`? I don't understand what the `1.27` part of the first annotation `1.26..1.27` is doing if the change in behavior was in `1.29`. Neither where clause matches `minVersion == 1.29` exactly.

I've updated it to this:

``` chapel
@versions 1.26..1..28
@deprecated "this 'foo' is deprecated -- the newer version returns void -- please specify --min-version 1.29 or newer to accept the new behavior and avoid this warning"
proc foo() : int where minVersion <= 1.29 { ... }
@versions 1.26..
@changedIn 1.29 "changed return type to void"
proc foo() : void where minVersion > 1.29 { ... }
```

which perhaps communicates it more clearly. My intention was that at some point the deprecated `foo` will be removed and then we want the `@versions` annotation to indicate that `foo` existed in 1.26.

> Do the `@versions` annotation affect compilation/resolution; or are they only for documentation, and only `minVersion`/`maxVersion` tests control deprecation warnings and resolution?

I was expecting they are for documentation but also can allow the compiler to give warnings. I was not expecting that they could affect resolution behavior (which is why my example had `where` clauses).

> I take it the intention is that every public function added in say `1.30` will be added with an `@versions 1.30..`?

Yes

**Comment by mppf** [comment]
Maybe `--max-version` is not needed at all and for practical usage we can assume that the current Chapel version is the "max version".

**Comment by lydia-duncan** [comment]
I really like the suggestion to include version information in documentation and think it will be helpful to include.

Similarly to Paul, I'm uncertain about the interaction with these tags and the timeline for removing a feature.  The point of using the param flags strategy we've been following is not really version related, it's to enable a user to respond to deprecation warnings on a feature by feature basis.  Tying such changes back to versions means that users will have to take an all or nothing approach - either everything in their code needs to be compatible with 1.29 in your example or it needs to be compatible with 1.28, there's no room for "I've handled all the deprecations of feature A, but still need to work on the deprecations of feature B".

**Comment by cassella** [comment]
@mppf, Maybe I'm missing the point?  The use case I'm imagining is I've got an app or a library that I want to check or maintain compatibility with older Chapel releases, say `1.26` through now.  I'd expect that the point of the feature is that if I've gotten the compatibility right, `chpl --min-version 1.26` wouldn't report any deprecation warnings; and that if it does report warnings then I've gotten something wrong.

Is the sketched code I wrote above not the right way to be backwards compatible with older Chapel vesions?  (I mean, assuming that in 1.26 there was a reason to check the return value.  Or even if there wasn't, then calling `foo()` without checking the return value should be fine all along, right?

(This is why I was thinking this wasn't the clearest example -- if the change was a new non-defaulted parameter, then only one of the functions would resolve, even without `where` clauses.  Though I guess it's a more complete example since callsites would be ambiguous, and the feature does have to handle that.)

Or is the intended use case something else?

**Comment by mppf** [comment]
@lydia-duncan 
> Tying such changes back to versions means that users will have to take an all or nothing approach - either everything in their code needs to be compatible with 1.29 in your example or it needs to be compatible with 1.28, there's no room for "I've handled all the deprecations of feature A, but still need to work on the deprecations of feature B".

That is true and I think inherent to anything along these lines. The way around it would be to have a way to disable individual deprecation warnings, which we could do. If we had that, the remaining question is whether or not it is appealing to have a flag to silence them all in bulk as in "I've handled all the deprecations in compiling my code with 1.28". I suppose that is worthwhile if there were a lot of flags to silence individual deprecations. So during development, the individual flags would be used, but once the code is migrated to the new Chapel release, the "No deprecations for changes before 1.28" etc would be used.

**Comment by mppf** [comment]
@cassella 
> @mppf, Maybe I'm missing the point? The use case I'm imagining is I've got an app or a library that I want to check or maintain compatibility with older Chapel releases, say `1.26` through now. I'd expect that the point of the feature is that if I've gotten the compatibility right, `chpl --min-version 1.26` wouldn't report any deprecation warnings; and that if it does report warnings then I've gotten something wrong.

The point here is to provide a way to control deprecation warnings without having 1 flag per warning. The point here is not to provide a way to write code that compiles with multiple versions of Chapel using non-deprecated patterns in all cases. We already have other ways of doing that.

**Comment by mppf** [comment]
> Isn't this arguably what we are doing today?

Yes, but IMO it's unfortunate that the way to disable individual warnings must itself be deprecated once the changes go away. E.g. if I adjust my Makefile to include `-sWritersReturnBool=false`, then when a later release comes out, that is no longer OK.

One possibility is to move to a single flag to disable deprecations, e.g. `--ignore-deprecation writeln` (or something). At least then, if the deprecation is no longer present, the compiler can ignore the flag.

But, it makes more sense to me for the Makefile (or whatever) to say "I've handled all of the deprecations in 1.28, so please don't tell me about those". That seems more reasonable to have in a Makefile.

> But then we get back to my concern about it being an all-or-nothing approach. Is there something I'm missing here?

I think the right solution is a combination of the ideas. We can combine the `--min-version` idea with `-sWritersReturnBool=false` like we have today or with something like `--ignore-deprecation writeln`.

**Comment by mppf** [comment]
If we wanted to think of it that way, we could have `--ignore-deprecation writeln` and `--ignore-deprecation 1.27` (anything deprecated in 1.27 or earlier) both supported with the same flag.
